File: demos_code_J1_conception_objet/03_lsp_rectangle_carre/DemoLspCorrige.java
/**
 * Une correction possible : ne pas promettre les setters de Rectangle.
 * Rectangle et Carre sont deux formes immuables partageant une capacité commune.
 */
public class DemoLspCorrige {
    interface Forme {
        int aire();
    }

    record Rectangle(int largeur, int hauteur) implements Forme {
        public Rectangle {
            if (largeur <= 0 || hauteur <= 0) {
                throw new IllegalArgumentException("Dimensions invalides");
            }
        }

        @Override
        public int aire() {
            return largeur * hauteur;
        }
    }

    record Carre(int cote) implements Forme {
        public Carre {
            if (cote <= 0) {
                throw new IllegalArgumentException("Côté invalide");
            }
        }

        @Override
        public int aire() {
            return cote * cote;
        }
    }

    static void afficherAire(Forme f) {
        System.out.println(f.getClass().getSimpleName() + " : aire = " + f.aire());
    }

    public static void main(String[] args) {
        afficherAire(new Rectangle(10, 5));
        afficherAire(new Carre(5));
    }
}

File: Demo en Python/demos_code_J1_conception_objet_python/03_lsp_rectangle_carre/DemoLspCorrige.py
from abc import ABC, abstractmethod
from dataclasses import dataclass


class Forme(ABC):
    @abstractmethod
    def aire(self) -> int:
        pass


@dataclass(frozen=True)
class Rectangle(Forme):
    largeur: int
    hauteur: int

    def __post_init__(self) -> None:
        if self.largeur <= 0 or self.hauteur <= 0:
            raise ValueError("Dimensions invalides")

    def aire(self) -> int:
        return self.largeur * self.hauteur


@dataclass(frozen=True)
class Carre(Forme):
    cote: int

    def __post_init__(self) -> None:
        if self.cote <= 0:
            raise ValueError("Côté invalide")

    def aire(self) -> int:
        return self.cote * self.cote


def afficher_aire(f: Forme) -> None:
    print(f"{type(f).__name__} : aire = {f.aire()}")


def main() -> None:
    afficher_aire(Rectangle(10, 5))
    afficher_aire(Carre(5))


if __name__ == "__main__":
    main()
